### PR TITLE
HW2

### PR DESCRIPTION
**Pull request by amauboussin**

### CODE REVIEW

File: HW2/P2/P2.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+Another note - I just acquired locks before accessing individual elements of the data rather than locking the decremented and incremented element at once. This means that my transactions are no longer atomic (it's possible for only half of the transaction to happen if the program is halted). However, this prevents some blocking and it will always turn out correct in the end if the program terminates. 
+
+Full results: 
+
+Serial uncorrelated: 0.301871061325 seconds
+Fine grained uncorrelated: 5.13503217697 seconds
+Medium grained uncorrelated (N=2): 6.20680999756 seconds
+Medium grained uncorrelated (N=5): 6.23391604424 seconds
+Medium grained uncorrelated (N=10): 6.29745292664 seconds
+Medium grained uncorrelated (N=50): 7.0954709053 seconds
```
Comment (zelong430) [inline HW2/P2/P2.txt:16]: Nice and Comprehensive testing. Great explanation!

File: HW2/P3/mandelbrot.pyx [hunk 2]
```diff
@@ -32,16 +32,66 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
...
+                    in_coords[i, j].imag
+                )
+
+                z_real = AVX.float_to_float8(0.0)
+                z_imaginary = AVX.float_to_float8(0.0)
+
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+                    #  0 if magnitude of z is greater than four, -1 otherwise
+                    mask = AVX.greater_than(
```
Comment (zelong430) [inline HW2/P3/mandelbrot.pyx:71]: clever way of masking on instruction level. Great Job!

File: HW2/P4/P4.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+In other words, a row's lower semaphore keeps track of how many iterations in comparison to its lower neighbor. Since the semaphore is never allowed to drop below 0 the neighbor cannot iterate past the row. 
+
+3.47082209587 seconds for 10 filter passes with 1 threads.
+1.77308177948 seconds for 10 filter passes with 2 threads.
+1.00484704971 seconds for 10 filter passes with 4 threads.
+0.925444126129 seconds for 10 filter passes with 6 threads.
+0.821120023727 seconds for 10 filter passes with 8 threads.
+0.798367977142 seconds for 10 filter passes with 16 threads.
+0.800353050232 seconds for 10 filter passes with 32 threads.
+0.806330919266 seconds for 10 filter passes with 64 threads.
```
Comment (zelong430) [inline HW2/P4/P4.txt:15]: I got similar results as yours and I think you grant a very good discussion about parallelization.

File: HW2/P5/P5.txt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+Using the grid method, our simulation only has to check the seven adjacent grid cells for possible collisions (O7N)=O(N) rather than checking every other ball (ON^2). Multithreading sped up both the grid and non-grid version by around 2.5 to 3x. The constant factor makes sense since neither the original version nor the grid version have any locks or other mechanisms that would block threads. 
+
+Sorting
+
+Sorting by Morton code further increased performance to 750 fps with one thread and 2100 fps with four threads. 
+This makes sense because the sorting ensures that balls close to each other on the grid will also be close to each other in memory. 
+
+Locking 
+
+Adding locks decreased performance to around 450 fps with one thread and 1250 fps with four threads. Since the locks decreased performance by roughly the same percentage in the single threaded and multithreaded case, it appears that most of the slowdown is caused by overhead rather than blocking.
```
Comment (zelong430) [inline HW2/P5/P5.txt:18]: Agree. The overhead of locking may be the main cause for the performance dropping.

File: HW2/P2/parallel_vector.pyx [hunk 5]
```diff
@@ -107,9 +112,14 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
...
                if counts[src[idx]] > 0:
+                   acquire(&(locks[dest[idx]/N]))
                    counts[dest[idx]] += 1
+                   release(&(locks[dest[idx]/N]))
+
+                   acquire(&(locks[src[idx]/N]))
                    counts[src[idx]] -= 1
+                   release(&(locks[src[idx]/N]))
 
    free_N_locks(num_locks, locks)
```
Comment (georgeqi94) [inline HW2/P2/parallel_vector.pyx]: This is how I did it too! :+1:

File: HW2/P2/parallel_vector.pyx [hunk 6]
```diff
@@ -107,9 +112,14 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
...
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
                if counts[src[idx]] > 0:
+                   acquire(&(locks[dest[idx]/N]))
                    counts[dest[idx]] += 1
+                   release(&(locks[dest[idx]/N]))
+
+                   acquire(&(locks[src[idx]/N]))
                    counts[src[idx]] -= 1
+                   release(&(locks[src[idx]/N]))
```
Comment (george-qi) [inline HW2/P2/parallel_vector.pyx:123]: Nice, that's how I did it as well!

File: HW2/P2/P2.txt [hunk 7]
```diff
@@ -0,0 +1,29 @@
+The serial approach performed much better than the parallel approach in both cases. This makes sense because the overhead of using the locks takes more time than simply moving the data.
+
+In the uncorrelated case the fine-grained locks performed better than any of coarse grained ones. In other words, no matter how few pieces of data were assigned to a lock, having less locks always reduced performance. This is a reasonable result because having more data per lock (a higher N) leads to more blocking. With uncorrelated data it makes sense to make N as small as possible (1)
+
+However, in the correlated case, the coarse-grained strategy can outperform the fine-grained. This is most liekly the result of many of the transactions only needing one lock (due to the correlation). When a transaction only needs one lock, there is less overhead but the chance of blocking does not substantially increase. The empirical results show that it's ideal to have ~20 pieces of data for each lock (N=20). 
+
```
Comment (george-qi) [inline HW2/P2/P2.txt:6]: ~20 sounds good! I think it would be good to note the theoretical reason for this result.

File: HW2/P2/P2.py [hunk 8]
```diff
@@ -69,14 +77,41 @@
...
+    plt.show()
+
+    keys, values = zip(*sorted(coarse_uncorrelated.items(), key = lambda x: x[0]))
+    values = [serial_uncorrelated, fine_grained_uncorrelated] + list(values)
+    labels = ['Serial', 'Fine Grained'] + ['N=%d' % n for n in keys]
+    positions = np.arange(len(values))
+    plt.bar(positions, values, align='center')
+    plt.xticks(positions, labels)
+    plt.ylabel('Time to move data')
+    plt.title('Uncorrelated Data Movement Times')
```
Comment (george-qi) [inline HW2/P2/P2.py:114]: Might just be easier to put the uncorrelated and correlated times on the same graph with 2 line graphs.

File: HW2/P3/P3.txt [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+With instruction-level parallelism:
+1 thread
+1487.42089435 million Complex FMAs / second
+2 threads
+3070.70750229 million Complex FMAs / second
+4 threads
+5957.29727686 million Complex FMAs / second
+
+
+Adding multithreading resulted in speed increases linear with the number of threads (up to four, this is reasonable since my machine has four cores and the problem is "embarassingly parallel"). Adding instruction level parallelism resulted in an approximately 8x increase over the first version - operations using the float8s were around as fast as operations with normal floats. Using the AVX interface is extra work but definitely pays off in terms of performance.
```
Comment (george-qi) [inline HW2/P3/P3.txt:22]: Nice -- I agree!

File: HW2/P3/mandelbrot.pyx [hunk 10]
```diff
@@ -32,16 +32,66 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
...
+                        c_real,
+                        AVX.sub(AVX.mul(z_real, z_real), AVX.mul(z_imaginary, z_imaginary))
+                    )
+                    z_imaginary = AVX.add(
+                        c_imaginary,
+                        AVX.add(AVX.mul(z_real, z_imaginary), AVX.mul(z_real, z_imaginary))
+                    )
+                    z_real = z_real_temp
+
+                AVX.to_mem(counts, &(out_counts[i, j]))
```
Comment (george-qi) [inline HW2/P3/mandelbrot.pyx:94]: Similar to how I did it as well!

File: HW2/P4/P4.txt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+My results shows performance increasing with the number of threads (although there are decreasing marginal returns to more threads) up to around 16. 
+
+
+My design makes each thread responsible for every n_threads row of the image. I allowed threads to begin asynchronously by creating two semaphores for each thread, one to regulate the slice above it and one to regulate the slice below it. These semaphores are acquired when an adjacent slice is trying to start filtering down a row and released when a slice finishes filtering a row. This approach ensures correctness and prevents deadlock because the semaphores prevent any one row from processed before its adjacencies - the total number of times the upper and lower semaphores have been released correspond to a row's progress, and the number of times they have been acquired correspond to the progress of neighboring rows. 
+
```
Comment (george-qi) [inline HW2/P4/P4.txt:5]: Neat way to do it and solid results.

File: HW2/P5/P5.txt (hunk 4 above)
Comment (george-qi) [inline HW2/P5/P5.txt:18]: Right -- I also got that locking made the performance drop significantly. Rest looks good!

File: HW2/P2/parallel_vector.pyx [hunk 12]
```diff
@@ -107,9 +112,14 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
                if counts[src[idx]] > 0:
```
Comment (zezhouliu) [inline HW2/P2/parallel_vector.pyx:116]: what if counts[src[idx]] changes between now and acquiring the lock? You usually want to check the same condition again after acquiring the lock

**Pull request closed without merging**